onnxscript/ir/traversal.py

```diff
@@ -0,0 +1,99 @@
+"""Utilities for traversing the IR graph."""
+
+from __future__ import annotations
+
+__all__ = [
+    "RecursiveGraphIterator",
+]
+
+from typing import Callable, Iterator, Reversible
+
+from typing_extensions import Self
+
+from onnxscript.ir import _core, _enums
+
+
+class RecursiveGraphIterator(Iterator[_core.Node], Reversible[_core.Node]):
+    def __init__(
+        self,
+        graph: _core.Graph | _core.Function | _core.GraphView,
+        *,
+        enter_graph_handler: Callable[[_core.Graph | _core.Function | _core.GraphView], None]
```

Comment:
I am unclear about the pros/cons of some of the design aspects here. 
* I can see the value in a utility that allows users to write ` for node in ir.traversal.RecursiveGraphIterator(model.graph)`.
* But, the moment users start providing enter_graph / exit_graph handlers, I personally feel that the advantage shifts to a design where the user defines a subclass (say, of RecursiveGraphIterator, which seems to be a Visitor pattern of some kind) by over-riding enter_graph / exit_graph methods. So, I would omit those parameters (in favor of user-defined sub-classes ... which may still be generic enough to be used by many, encoding a specific style of visiting).

Reply:
I think that the idea of decoupling `Pass` from a `GraphIterator\Visitor` makes sense. They are two separate infrastructure components.

Reply:
That works too. Do you envision the following: to record the current scope in the iteration, one would do:

```python
class ScopeRecursiveGraphIterator(RecursiveGraphIterator):
    def __init__(self, scope):
        self.scope = []

	def enter_graph_handler(graph):
	    scopes.append(graph.name)

	def exit_graph_handler(graph):
	    scopes.pop()

scope = []
for node in ScopeRecursiveGraphIterator(graph, scope):
    print(scopes)
```

instead of

```python
scopes = []

def enter_graph_handler(graph):
    scopes.append(graph.name)

def exit_graph_handler(graph):
    scopes.pop()

for node in traversal.RecursiveGraphIterator(
    graph, enter_graph_handler=enter_graph_handler, exit_graph_handler=exit_graph_handler
):
    print(scopes)
```

?

Reply:
A challenge is that in the recursive calls, we don't know how to initialize the user defined classes.

Reply:
Perhaps providing just the basic GraphIterator (without the ability to specify enter/exit graph handlers) would be a reasonable starting point? As we complete the other applications, such as migrating the Optimizer etc., we may find what extensions we need.

Reply:
Done